numba/unicode.py

```diff
@@ -1286,6 +1288,134 @@ def impl(a):
+    mixed with:
+    https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Objects/bytes_methods.c#L131-L156
+    """
+
+    def impl(data):
+        length = len(data)
+        if length == 1:
+            return _PyUnicode_IsLowercase(_get_code_point(data, 0))
+        if length == 0:
+            return False
+
+        cased = False
+        for idx in range(length):
+            code_point = _get_code_point(data, idx)
+            if _PyUnicode_IsUppercase(code_point) or _PyUnicode_IsTitlecase(code_point):
+                return False
+            elif not cased and _PyUnicode_IsLowercase(code_point):
+                cased = True
+        return cased
+    return impl
+
+
+@register_jitable
+def _handle_capital_sigma(data, length, idx):
+    """
+    Handle capital sigma
+    https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Objects/unicodeobject.c#L9856-L9883
+    """
+    c = 0
+    j = idx - 1
+    while j >= 0:
+        c = _get_code_point(data, j)
+        if not _PyUnicode_IsCaseIgnorable(c):
+            break
+        j -= 1
+    final_sigma = (j >= 0 and _PyUnicode_IsCased(c))
+    if final_sigma:
+        j = idx + 1
+        while j < length:
+            c = _get_code_point(data, j)
+            if not _PyUnicode_IsCaseIgnorable(c):
+                break
+            j += 1
+        final_sigma = (j == length or (not _PyUnicode_IsCased(c)))
+    return 0x3c2 if final_sigma else 0x3c3
+
+
+@register_jitable
+def _lower_ucs4(code_point, data, length, idx, mapped):
+    """
+    https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Objects/unicodeobject.c#L9885-L9895
+    """
+    if code_point == 0x3A3:
+        mapped[0] = _handle_capital_sigma(data, length, idx)
+        return 1
+    return _PyUnicode_ToLowerFull(code_point, mapped)
+
+
+@register_jitable
+def _do_upper_or_lower(data, length, res, maxchars, lower):
+    """
+    https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Objects/unicodeobject.c#L9946-L9965
+    """
+    k = 0
+    for idx in range(length):
+        mapped = np.zeros(3, dtype=_Py_UCS4)
+        code_point = _get_code_point(data, idx)
+        if lower:
+            n_res = _lower_ucs4(code_point, data, length, idx, mapped)
+        else:
+            n_res = _PyUnicode_ToUpperFull(code_point, mapped)
+        for m in mapped[:n_res]:
+            maxchars[0] = max(maxchars[0], m)
+            _set_code_point(res, k, m)
+            k += 1
+    return k
+
+
+@register_jitable
+def _do_lower(data, length, res, maxchars):
+    """
+    https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Objects/unicodeobject.c#L9973-L9977
+    """
+    return _do_upper_or_lower(data, length, res, maxchars, lower=True)
+
+
+@overload_method(types.UnicodeType, 'lower')
+def unicode_lower(data):
```

Comment:
This is so similar to `unicode_upper` it ought to just be generated instead of this duplication?!

Reply:
I think `unicode_lower` and `unicode_upper` should be refactored in separate commit. Current commit covers only `lower/islower`. I can't implement `lower` using existing functionality and not changing `upper` method.

Reply:
that seems fine to do, thanks.